### DISCUSSION TITLE
gptel-rewrite often fails to generate code-only responses

### DISCUSSION THREAD

**Discussion by guibor** [karthink/gptel#537] [Q&A]
### Please update gptel first -- errors are often fixed by the time they're reported.

- [X] I have updated gptel to the latest commit and tested that the issue still exists

### Bug Description

When using gptel-rewrite on a region, I get full responses with explanations (e.g., "Install package X, and then do this" rather than a diff that can be applied)

### Backend

None

### Steps to Reproduce

Any gptel-rewrite on Python code often reproduces this.

### Additional Context

29.2, Mac OS X

**Comment by karthink** [comment]
1. Can you give me an example of the rewrite directive you supply?
2. What model are you using?

**Comment by guibor** [comment]
I don't have the directive but it wasn't the default one that instructs the model to only output code (I didn't realize that gptel-rewrite relies on the directive for that).
Might be worth making into an option (set by default to true in prog-mode buffers, or at least have all other words in comments)  rather than adding to a potentially long directive

**Comment by karthink** [comment]
Reopening since there is either a bug or some misunderstanding.

> (I didn't realize that gptel-rewrite relies on the directive for that).

It does not.  The only instruction fed to the rewrite request is whatever you give it in the menu:
![image](https://github.com/user-attachments/assets/b48cb9c2-b1e9-45e0-ab1d-86819626b8fe)

So you do not need to set the system message every time you want to call the rewrite function.

> Might be worth making into an option

So this doesn't apply.  However, I do want to make it easier to specify the rewrite instructions, by templating out the common part (the part that instructs the model to respond without comments).

----------

That brings us to your question: what rewrite instructions were you giving the model when it generated general text instead of code?

**Comment by karthink** [comment]
Moving to discussions as there is no bug here to fix.

**Comment by CiroDornelles** [comment]
This worked for me

You are a very helpful AI assistant. You are currently playing a game called "Silent Answers for Bash." The game works as follows: you receive any question, such as "How do I list my files in this folder?" and you must return a card to the user containing only the answer they seek, such as `ls`. You must not provide any extra information and should only return the card without saying anything else under any circumstances.
